File: tks/RepositoriesAdapters/src/main/kotlin/adapters/UserRepositoryAdapter.java
package adapters;

import data.UserEntity;
import drivenports.UserQueryPort;
import drivenports.UserSavePort;
import mappers.UserMapper;
import model.User;
import repositories.RepositoryBase;

import java.util.Objects;

public class UserRepositoryAdapter implements UserSavePort, UserQueryPort {

    private final RepositoryBase<UserEntity> repository;
    private final UserMapper mapper;

    public UserRepositoryAdapter(RepositoryBase<UserEntity> repository, UserMapper mapper) {
        this.repository = repository;
        this.mapper = mapper;
    }

    @Override
    public User findByEmail(String email) {
        var entity = repository.find(x -> x.getEmail().equals(email));
        return mapper.mapEntityToDomainObject(entity);
    }

    @Override
    public void add(User user) {
        var entity = mapper.mapDomainObjectToEntity(user);
        repository.add(entity);
    }

    @Override
    public void update(User user) {
        var entity = repository.find(x -> Objects.equals(x.getEmail(), user.getEmail()));
        assert entity != null; // todo prepare proper exception
        mapper.mapDomainObjectToEntity(user, entity);
        repository.update(entity);
    }
}

File: tks/RepositoriesAdapters/_src/main/kotlin/repository/adapters/UserRepositoryAdapter.kt
package repository.adapters

import domain.model.context.users.User
import domain.model.values.Email
import ports.secondary.IUserRepositoryAdapter
import repository.data.UserEntity
import repository.mappers.UserMapper
import repository.repositories.IRepository

class UserRepositoryAdapter(
    private val repository: IRepository<UserEntity>,
    private val mapper: UserMapper
) : IUserRepositoryAdapter {

    override fun findByEmail(email: Email): User? {
        val entity = repository.find { x: UserEntity -> x.email == email.value }
        return mapper.mapEntityToDomainObject(entity)
    }

    override fun save(user: User) {
        val entity = repository.find { x: UserEntity -> x.email == user.email.value }
        if (entity == null) {
            repository.add(mapper.mapDomainObjectToEntity(user)!!)
        } else {
            mapper.mapDomainObjectToEntity(user, entity)
            repository.update(entity)
        }
    }
}